paasta_tools/cli/cmds/mark_for_deployment.py

```diff
@@ -323,15 +366,57 @@ def instances_deployed(cluster, service, instances, git_sha):
                              .format(service, instance, cluster,
                                      status.marathon.running_instance_count,
                                      status.marathon.expected_instance_count))
+                instances_in.task_done()
+                instances_out.put(instance)
                 continue
             paasta_print("Complete: {}.{} on {} looks 100% deployed at {} "
                          "instances on {}"
                          .format(service, instance, cluster,
                                  status.marathon.running_instance_count,
                                  status.git_sha))
-            result.append(instance)
+            instances_in.task_done()
+
 
-    return result
+def _query_clusters(cluster_map, service, git_sha, green_light):
+    """Run _run_cluster_worker() in a separate thread for each paasta cluster
+
+    Args:
+        green_light: an instance of threading.Event(). It is used to terminate
+            all running threads after KeyboardInterrupt.
+    """
+    for cluster, cluster_data in cluster_map.items():
+        if cluster_data['deployed'] != cluster_data['num_instances']:
+            worker = Thread(target=_run_cluster_worker,
+                            args=(cluster, cluster_data, service, git_sha,
+                                  green_light))
+            worker.start()
+            cluster_data['worker'] = worker
```

Comment:
@oktopuz Hm, you're right, maybe this is ok if you're only using the dict in the scope of this function. 
But here's why I'm generally against this:

Consider this:

```python

def foo(bar):
   bar['y'] = 'hi'

def baz(bar):
   print bar['y']

def qux(bar):
   print bar['x']
```

We use the common arg ``bar`` everywhere, but it means different things in different places. ``foo`` requires just a dict, ``baz`` requires a dict with the key ``y``, and ``qux`` requires a dict with the key ``x``. This doesn't really get encoded anywhere though; the caller has to know the implementation details of the function to know how to give it the 'right' thing. If you make NamedTuples for ``foo`` ``baz`` and ``qux``, you put a contract around what they have to look like.

In our case, I'd really like ``cluster_map`` and ``cluster_data`` dicts to be replaced with ``NamedTuple``. Then, rather than mutating it, you create new data structures that encode the coupling of ``cluster_map`` or ``cluster_data`` and some other thing (the worker, in this case). At the same time, I get that this wasn't your doing, so I don't expect you to refactor that. But, as a halfway house, I'd do this:

```python
workers = []
for cluster, cluster_data in cluster_map.items():
    if cluster_data['deployed'] != len(cluster_data['instances']):
             worker = Thread(target=_run_cluster_worker,
                             args=(cluster, cluster_data, service, git_sha,
                                   green_light))
             worker.start()
             workers.append(worker)   
            
for worker in workers:
    try:
        while green_light.is_set() and worker.isAlive():
            time.sleep(.2)
    except (KeyboardInterrupt, SystemExit):
        green_light.clear()
        paasta_print('KeyboardInterrupt received. Terminating..')
        cluster_data['worker'].join()
```

As far as I can tell, the second for loop doesn't the cluster_map for anything other than the worker, so why not just store each of the workers in a separate list?